python/pyspark/sql/conversion.py

```diff
@@ -0,0 +1,543 @@
+
+            def convert_other(value: Any) -> Any:
+                if value is None:
+                    raise PySparkValueError(f"input for {dataType} must not be None")
+                return value
+
+            return convert_other
+        else:
+            return lambda value: value
+
+    @staticmethod
+    def convert(data: Sequence[Any], schema: StructType, use_large_var_types: bool) -> "pa.Table":
+        assert isinstance(data, list) and len(data) > 0
+
+        assert schema is not None and isinstance(schema, StructType)
+
+        column_names = schema.fieldNames()
+
+        column_convs = [
+            LocalDataToArrowConversion._create_converter(
+                field.dataType, field.nullable, variants_as_dicts=True
+            )
+            for field in schema.fields
+        ]
+
+        pylist: List[List] = [[] for _ in range(len(column_names))]
+
+        for item in data:
+            if isinstance(item, VariantVal):
+                raise PySparkValueError("Rows cannot be of type VariantVal")
+            if (
+                not isinstance(item, Row)
+                and not isinstance(item, tuple)  # inherited namedtuple
+                and hasattr(item, "__dict__")
+            ):
+                item = item.__dict__
+            if isinstance(item, dict):
+                for i, col in enumerate(column_names):
+                    pylist[i].append(column_convs[i](item.get(col)))
+            else:
+                if len(item) != len(column_names):
+                    raise PySparkValueError(
+                        errorClass="AXIS_LENGTH_MISMATCH",
+                        messageParameters={
+                            "expected_length": str(len(column_names)),
+                            "actual_length": str(len(item)),
+                        },
+                    )
+
+                for i in range(len(column_names)):
+                    pylist[i].append(column_convs[i](item[i]))
+
+        pa_schema = to_arrow_schema(
+            StructType(
+                [
+                    StructField(
+                        field.name, _deduplicate_field_names(field.dataType), field.nullable
+                    )
+                    for field in schema.fields
+                ]
+            ),
+            prefers_large_types=use_large_var_types,
+        )
+
+        return pa.Table.from_arrays(pylist, schema=pa_schema)
+
+
+class ArrowTableToRowsConversion:
+    """
+    Conversion from Arrow Table to Rows.
+    """
+
+    @staticmethod
+    def _need_converter(dataType: DataType) -> bool:
+        if isinstance(dataType, NullType):
+            return True
+        elif isinstance(dataType, StructType):
+            return True
+        elif isinstance(dataType, ArrayType):
+            return ArrowTableToRowsConversion._need_converter(dataType.elementType)
+        elif isinstance(dataType, MapType):
+            # Different from PySpark, here always needs conversion,
```

Comment:
I remember there is another version of conversion for PySpark classic, do we plan to unify them in the future?

Reply:
Are you referring to [`pyspark.sql.types._create_converter`](https://src.dev.databricks.com/databricks-eng/runtime@72929e1baf2ed733e300437b784ca4f2e32f3d7a/-/blob/python/pyspark/sql/types.py?L2567-2617)? Seems like there are some differences, for example `ArrowTableToRowsConversion` handles UDT and variant but the one in `types.py` doesn't.

Reply:
Can we check where `_create_converter`is used and see if we can merge these two methods?

Reply:
actually, I meant `python/pyspark/sql/pandas/conversion.py`. We can investigate it later.